# **ТИПиС - ИДЗ№1 - Воропаев Илья 3391**

In [ ]:
print('Hello World!')

Hello World!


In [2]:
import pandas as pd

In [3]:
pip install ucimlrepo

In [4]:
from ucimlrepo import fetch_ucirepo
# загрузка dataset
adult = fetch_ucirepo(id=2)

In [6]:
print(adult.data.features)

       age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48837   39           Private  215419  Bachelors             13   
48838   64               NaN  321403    HS-grad              9   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation    relationship  \
0           Never-married       Adm-clerical   Not-in-family   
1      Married-civ-spouse    Exec-managerial         Husband   
2              

In [7]:
print(adult.data.targets)

       income
0       <=50K
1       <=50K
2       <=50K
3       <=50K
4       <=50K
...       ...
48837  <=50K.
48838  <=50K.
48839  <=50K.
48840  <=50K.
48841   >50K.

[48842 rows x 1 columns]


In [8]:
# объединение фреймов по столбцам
data = pd.concat([adult.data.features, adult.data.targets], axis=1)

In [9]:
data

age         workclass  fnlwgt  education  education-num  \
0       39         State-gov   77516  Bachelors             13   
1       50  Self-emp-not-inc   83311  Bachelors             13   
2       38           Private  215646    HS-grad              9   
3       53           Private  234721       11th              7   
4       28           Private  338409  Bachelors             13   
...    ...               ...     ...        ...            ...   
48837   39           Private  215419  Bachelors             13   
48838   64               NaN  321403    HS-grad              9   
48839   38           Private  374983  Bachelors             13   
48840   44           Private   83891  Bachelors             13   
48841   35      Self-emp-inc  182148  Bachelors             13   

           marital-status         occupation    relationship  \
0           Never-married       Adm-clerical   Not-in-family   
1      Married-civ-spouse    Exec-managerial         Husband   
2                Divorced  Handlers-cleaners   Not-in-family   
3      Married-civ-spouse  Handlers-cleaners         Husband   
4      Married-civ-spouse     Prof-specialty            Wife   
...                   ...                ...             ...   
48837            Divorced     Prof-specialty   Not-in-family   
48838             Widowed                NaN  Other-relative   
48839  Married-civ-spouse     Prof-specialty         Husband   
48840            Divorced       Adm-clerical       Own-child   
48841  Married-civ-spouse    Exec-managerial         Husband   

                     race     sex  capital-gain  capital-loss  hours-per-week  \
0                   White    Male          2174             0              40   
1                   White    Male             0             0              13   
2                   White    Male             0             0              40   
3                   Black    Male             0             0              40   
4                   Black  Female             0             0              40   
...                   ...     ...           ...           ...             ...   
48837               White  Female             0             0              36   
48838               Black    Male             0             0              40   
48839               White    Male             0             0              50   
48840  Asian-Pac-Islander    Male          5455             0              40   
48841               White    Male             0             0              60   

      native-country  income  
0      United-States   <=50K  
1      United-States   <=50K  
2      United-States   <=50K  
3      United-States   <=50K  
4               Cuba   <=50K  
...              ...     ...  
48837  United-States  <=50K.  
48838  United-States  <=50K.  
48839  United-States  <=50K.  
48840  United-States  <=50K.  
48841  United-States   >50K.  

[48842 rows x 15 columns]

In [10]:
len(data.columns) # кол-во столбцов в таблице (хотя уже понятно, что 15)

15

In [11]:
len(data.index) # кол-во строк в таблице = кол-во индексов

48842

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       47879 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      47876 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48568 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


Всего строк 48842. Значит надо искать столбцы, где ненулевых строк меньше. Таковыми являются workclass, occupation, native-country.

In [13]:
len(data['race'].unique()) # кол-во уникальных значений в столбце race

5

In [15]:
data['hours-per-week'].median() # медиана hours-per-week

40.0

In [19]:
high_income = data[data['income'] == '>50K'] # фильтруем данные только с зарплатой >50K
gender_count = high_income['sex'].value_counts() # считаем количество мужчин и женщин в отфильтрованных данных

In [21]:
len(high_income)

7841

In [20]:
gender_count

sex
Male      6662
Female    1179
Name: count, dtype: int64

Мужчин с зарплатой >50К больше, чем женщин.

In [25]:
# наиболее встречаемые значения в столбцах с пропусками
often_value1 = data['workclass'].mode()
often_value2 = data['occupation'].mode()
often_value3 = data['native-country'].mode()

In [34]:
often_value3

0    United-States
Name: native-country, dtype: object

Моды может быть две, поэтому берём первую попавшуюся.

In [30]:
# изменяем фрейм напрямую
data['workclass'].fillna(data['workclass'].mode()[0], inplace = True)
data['occupation'].fillna(data['occupation'].mode()[0], inplace = True)
data['native-country'].fillna(data['native-country'].mode()[0], inplace = True)

/tmp/ipython-input-2090364661.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['occupation'].fillna(data['occupation'].mode()[0], inplace = True)
/tmp/ipython-input-2090364661.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value,

In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       48842 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      48842 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48842 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


# **Ответы на вопросы:**

1. Количество столбцов в наборе данных - 15.
2. Пропуски есть в столбцах: workclass, occupation, native-country.
3. Количество уникальных значений в столбце race - 5.
4. Медиана hours-per-week - 40.0.
5. Мужчин с зарплатой >50К больше, чем женщин.
6. Пропуски были заполнены наиболее встречаемыми значениями. Так как это категориальные данные, то пропуски можно было заполнить дефолтными значениями вроде 'lost' или 'not defined'.